File: server/src/main/java/io/spine/server/entity/EntityVisibility.java
/*
 * Copyright 2019, TeamDev. All rights reserved.
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.server.entity;

import com.google.common.graph.GraphBuilder;
import com.google.common.graph.ImmutableGraph;
import com.google.common.graph.MutableGraph;
import com.google.errorprone.annotations.Immutable;
import com.google.protobuf.Descriptors.Descriptor;
import com.google.protobuf.Message;
import io.spine.annotation.Internal;
import io.spine.option.EntityOption;
import io.spine.option.EntityOption.Kind;
import io.spine.option.EntityOption.Visibility;
import io.spine.option.OptionsProto;
import io.spine.type.TypeName;

import java.io.Serializable;

import static com.google.common.base.Preconditions.checkArgument;
import static com.google.common.base.Preconditions.checkNotNull;
import static io.spine.option.EntityOption.Kind.PROJECTION;
import static io.spine.option.EntityOption.Visibility.DEFAULT;
import static io.spine.option.EntityOption.Visibility.FULL;
import static io.spine.option.EntityOption.Visibility.NONE;
import static io.spine.option.EntityOption.Visibility.QUERY;
import static io.spine.option.EntityOption.Visibility.SUBSCRIBE;

/**
 * The visibility of an entity type.
 *
 * <p>An entity can have one of the following visibility levels:
 * <ol>
 *     <li>{@code NONE} - the entity is not visible to the clients for reading;
 *     <li>{@code QUERY} - the entity is visible for querying, but not for subscription;
 *     <li>{@code SUBSCRIBE} - the entity is visible for subscription, but not for querying;
 *     <li>{@code FULL} - the entity is visible for both subscription and querying.
 * </ol>
 *
 * <p>The visibility of an entity is defined by the {@code (entity)} option. By default, any entity
 * is has the {@code NONE} level except for projections, which have the {@code FULL} level.
 */
@Immutable
@Internal
public final class EntityVisibility implements Serializable {

    private static final long serialVersionUID = 0L;

    /**
     * An ordered immutable graph of the visibility levels.
     *
     * <p>This graph is used for checking if a certain level includes another level. For example,
     * {@code FULL} includes {@code QUERY}. To represent this, there is an edge from {@code FULL}
     * to {@code QUERY} in the graph.
     *
     * @see #isAsLeast(Visibility)
     */
    private static final ImmutableGraph<Visibility> VISIBILITIES = buildVisibilityGraph();

    private final Visibility value;

    private EntityVisibility(Visibility value) {
        this.value = value;
    }

    /**
     * Obtains an instance of {@code EntityVisibility} for the given entity state class.
     *
     * @param stateClass
     *         the entity state class
     * @return new instance
     */
    public static EntityVisibility of(Class<? extends Message> stateClass) {
        checkNotNull(stateClass);
        Descriptor descriptor = TypeName.of(stateClass)
                                        .messageDescriptor();
        EntityOption entityOption = descriptor.getOptions()
                                              .getExtension(OptionsProto.entity);
        Visibility visibility = entityOption.getVisibility();
        if (visibility == DEFAULT) {
            Kind kind = entityOption.getKind();
            visibility = kind == PROJECTION ? FULL : NONE;
        }
        return new EntityVisibility(visibility);
    }

    /**
     * Checks if the visibility level is sufficient for subscription.
     *
     * <p>If the level is sufficient, it is allowed to subscribe to the entity state updates via
     * {@link io.spine.core.Subscribe} or {@link io.spine.server.SubscriptionService}.
     *
     * @return {@code true} if the visibility is {@code SUBSCRIBE} or {@code FULL}, {@code false}
     *         otherwise
     */
    public boolean canSubscribe() {
        return isAsLeast(SUBSCRIBE);
    }

    /**
     * Checks if the visibility level is sufficient for querying.
     *
     * <p>If the level is sufficient, it is allowed to query the entity states via
     * {@link io.spine.server.QueryService}.
     *
     * @return {@code true} if the visibility is {@code QUERY} or {@code FULL}, {@code false}
     *         otherwise
     */
    public boolean canQuery() {
        return isAsLeast(QUERY);
    }

    /**
     * Checks if the visibility level is not {@code NONE}.
     *
     * @return {@code true} if the visibility is anything but {@code NONE}, {@code false} otherwise
     */
    public boolean isNotNone() {
        return value != NONE;
    }

    /**
     * Checks if the visibility is exactly of the given level.
     *
     * @param visibility
     *         the visibility to compare to
     * @return {@code true} if this visibility is equal to the given visibility
     * @throws IllegalArgumentException
     *         if the given visibility is {@code DEFAULT}
     */
    public boolean is(Visibility visibility) {
        checkNotNull(visibility);
        checkNotDefault(visibility);
        return value == visibility;
    }

    /**
     * Checks if the visibility is at least as allowing as the given one.
     *
     * <p>{@code isAsLeast(NONE)} always returns {@code true}. {@code isAtLeast(FULL)} returns
     * {@code true} only if this visibility is exactly {@code FULL}.
     *
     * @param visibility
     *         the visibility to compare to
     * @return {@code true} if this visibility is equal to the given visibility
     * @throws IllegalArgumentException
     *         if the given visibility is {@code DEFAULT}
     */
    public boolean isAsLeast(Visibility visibility) {
        checkNotNull(visibility);
        checkNotDefault(visibility);
        return VISIBILITIES.hasEdgeConnecting(value, visibility);
    }

    private static void checkNotDefault(Visibility visibility) {
        checkArgument(visibility != DEFAULT, "Unexpected visibility level: %s.", DEFAULT);
    }

    @Override
    public String toString() {
        return value.name();
    }

    /**
     * Builds the graph of the visibilities hierarchy.
     *
     * @see #VISIBILITIES
     */
    private static ImmutableGraph<Visibility> buildVisibilityGraph() {
        MutableGraph<Visibility> mutableGraph = GraphBuilder
                .directed()
                .expectedNodeCount(Visibility.values().length)
                .allowsSelfLoops(true)
                .build();
        mutableGraph.addNode(NONE);
        mutableGraph.addNode(QUERY);
        mutableGraph.addNode(SUBSCRIBE);
        mutableGraph.addNode(FULL);

        mutableGraph.putEdge(QUERY, NONE);
        mutableGraph.putEdge(SUBSCRIBE, NONE);
        mutableGraph.putEdge(FULL, SUBSCRIBE);
        mutableGraph.putEdge(FULL, QUERY);
        mutableGraph.putEdge(FULL, NONE);

        mutableGraph.nodes()
                    .forEach(visibility -> mutableGraph.putEdge(visibility, visibility));
        return ImmutableGraph.copyOf(mutableGraph);
    }
}

File: server/src/main/kotlin/io/spine/server/entity/EntityVisibility.kt
/*
 * Copyright 2025, TeamDev. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Redistribution and use in source and/or binary forms, with or without
 * modification, must retain the above copyright notice and the following
 * disclaimer.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

package io.spine.server.entity

import com.google.common.graph.GraphBuilder
import com.google.common.graph.ImmutableGraph
import com.google.errorprone.annotations.Immutable
import io.spine.annotation.Internal
import io.spine.base.EntityState
import io.spine.code.proto.EntityStateOption
import io.spine.option.EntityOption.Kind.PROJECTION
import io.spine.option.EntityOption.Visibility
import io.spine.option.EntityOption.Visibility.DEFAULT
import io.spine.option.EntityOption.Visibility.FULL
import io.spine.option.EntityOption.Visibility.NONE
import io.spine.option.EntityOption.Visibility.QUERY
import io.spine.option.EntityOption.Visibility.SUBSCRIBE
import io.spine.server.entity.EntityVisibility.Companion.visibilities
import io.spine.type.TypeName
import java.io.Serial
import java.io.Serializable
import java.util.*

/**
 * The visibility of an entity type.
 *
 * An entity can have one of the following visibility levels:
 *
 *  1. [NONE] — the entity is not visible to the clients for reading;
 *  2. [QUERY] — the entity is visible for querying, but not for subscription;
 *  3. [SUBSCRIBE] — the entity is visible for subscription, but not for querying;
 *  4. [FULL] — the entity is visible for both subscription and querying.
 *
 * The visibility of an entity is defined by the `(entity)` option.
 * By default, any entity has the `NONE` level except for projections,
 * which have the `FULL` level.
 */
@Immutable
@Internal
public class EntityVisibility private constructor(
    private val value: Visibility
) : Serializable {

    /**
     * Checks if the visibility level is sufficient for subscription.
     *
     * If the level is sufficient, it is allowed to subscribe to the entity state updates via
     * [io.spine.core.Subscribe] or [io.spine.server.SubscriptionService].
     *
     * @return `true` if the visibility is `SUBSCRIBE` or `FULL`, `false` otherwise.
     */
    public fun canSubscribe(): Boolean = isAsLeast(SUBSCRIBE)

    /**
     * Checks if the visibility level is sufficient for querying.
     *
     * If the level is sufficient, it is allowed to query the entity states via
     * [io.spine.server.QueryService].
     *
     * @return `true` if the visibility is `QUERY` or `FULL`, `false` otherwise.
     */
    public fun canQuery(): Boolean = isAsLeast(QUERY)

    /**
     * Checks if the visibility level is not `NONE`.
     *
     * @return `true` if the visibility is anything but `NONE`, `false` otherwise
     */
    public fun isNotNone(): Boolean = value != NONE

    /**
     * Checks if the visibility is exactly of the given level.
     *
     * @param visibility The visibility to compare to.
     * @return `true` if this visibility is equal to the given visibility
     * @throws IllegalArgumentException if the given visibility is `DEFAULT`.
     */
    @Suppress("FunctionNaming") // Keep backward compatibility for Java code.
    public fun `is`(visibility: Visibility): Boolean {
        checkNotDefault(visibility)
        return value == visibility
    }

    /**
     * Checks if the visibility is at least as allowing as the given one.
     *
     * `isAsLeast(NONE)` always returns `true`.
     * `isAtLeast(FULL)` returns `true` only if this visibility is exactly `FULL`.
     *
     * @param visibility The visibility to compare to.
     * @return `true` if this visibility is equal to the given visibility.
     * @throws IllegalArgumentException if the given visibility is `DEFAULT`.
     */
    public fun isAsLeast(visibility: Visibility): Boolean {
        checkNotDefault(visibility)
        return visibilities.hasEdgeConnecting(value, visibility)
    }

    override fun toString(): String = value.name

    @Suppress("UnstableApiUsage") // `ImmutableGraph` is in `@Beta`.
    public companion object {

        @Serial
        private const val serialVersionUID = 0L

        /**
         * An ordered immutable graph of the visibility levels.
         *
         * This graph is used for checking if a certain level includes another level.
         * For example, `FULL` includes `QUERY`.
         * To represent this, there is an edge from `FULL` to `QUERY` in the graph.
         *
         * @see isAsLeast
         */
        private val visibilities: ImmutableGraph<Visibility> by lazy {
            buildVisibilityGraph()
        }

        /**
         * Obtains an instance of `EntityVisibility` for the given entity state class.
         *
         * Returns `Optional.empty()`, if the given message is not an `Entity`,
         * i.e., has no corresponding option defined.
         *
         * @param stateClass The entity state class.
         * @return new instance.
         */
        @JvmStatic
        public fun of(stateClass: Class<out EntityState<*>>): Optional<EntityVisibility> {
            val descriptor = TypeName.of(stateClass).messageDescriptor()
            val option = EntityStateOption.valueOf(descriptor)
            if (option.isEmpty) {
                return Optional.empty<EntityVisibility>()
            }
            val entityOption = option.get()
            var visibility = entityOption.getVisibility()
            if (visibility == DEFAULT) {
                val kind = entityOption.getKind()
                visibility =
                    if (kind == PROJECTION) FULL else NONE
            }
            val result = EntityVisibility(visibility)
            return Optional.of<EntityVisibility>(result)
        }

        private fun checkNotDefault(visibility: Visibility) =
            require(visibility != DEFAULT) {
                "Unexpected visibility level: ${DEFAULT}."
            }

        /**
         * Builds the graph of the visibility hierarchy.
         *
         * @see visibilities
         */
        private fun buildVisibilityGraph(): ImmutableGraph<Visibility> {
            val mutableGraph = GraphBuilder
                .directed()
                .expectedNodeCount(Visibility.entries.size)
                .allowsSelfLoops(true)
                .build<Visibility>()

            mutableGraph.run {
                addNode(NONE)
                addNode(QUERY)
                addNode(SUBSCRIBE)
                addNode(FULL)

                putEdge(QUERY, NONE)
                putEdge(SUBSCRIBE, NONE)
                putEdge(FULL, SUBSCRIBE)
                putEdge(FULL, QUERY)
                putEdge(FULL, NONE)
            }

            mutableGraph.nodes().forEach { visibility ->
                mutableGraph.putEdge(visibility, visibility)
            }
            return ImmutableGraph.copyOf<Visibility>(mutableGraph)
        }
    }
}
